src/audio/audio.cpp

```diff
@@ -125,28 +158,78 @@ Set the master output volume.
 
 @param[in] volume   the master volume (between 0 and 1)
 */
-void Audio::setOutputVolume(ALfloat volume)
+void Audio::setOutputVolume(qreal volume)
 {
     QMutexLocker locker(&audioLock);
 
-    volume = std::max(0.f, std::min(volume, 1.f));
+    volume = std::max(0.0, std::min(volume, 1.0));
 
-    alListenerf(AL_GAIN, volume);
+    alListenerf(AL_GAIN, static_cast<ALfloat>(volume));
     checkAlError();
 }
 
-ALfloat Audio::inputVolume()
+/**
+@brief The minimum gain value for an input device.
+
+@return minimum gain value in dB
+*/
+qreal Audio::minInputGain() const
 {
     QMutexLocker locker(&audioLock);
+    return d->minInputGain;
+}
+
+/**
+@brief Set the minimum allowed gain value in dB.
 
-    return inGain;
+@note Default is -30dB; usually you don't need to alter this value;
+*/
+void Audio::setMinInputGain(qreal dB)
+{
+    QMutexLocker locker(&audioLock);
+    d->minInputGain = dB;
 }
 
-void Audio::setInputVolume(ALfloat volume)
+/**
+@brief The maximum gain value for an input device.
+
+@return maximum gain value in dB
+*/
+qreal Audio::maxInputGain() const
 {
     QMutexLocker locker(&audioLock);
+    return d->maxInputGain;
+}
+
+/**
+@brief Set the maximum allowed gain value in dB.
 
-    inGain = std::max(0.f, std::min(volume, 1.f));
+@note Default is 30dB; usually you don't need to alter this value.
+*/
+void Audio::setMaxInputGain(qreal dB)
```

Comment:
>What is the advantage of creating a whole new class to encapsulate mutable parameters that that are essentially only changed by this class?

It is more than just that. Actually, I prepare to encapsulate the qTox Audio API completely and bring back the pimpl idiom. This is absolutely necessary, to keep the API controllable, making it "future proof" (had a hard time btw. collecting all the code fragments and clean up the mess to get to the current stable state). Further it actually eases handling of audio data within qTox by providing a sane interface, which cannot be mistaken. It has some other advantages as well. For example it allows an atomic replacement of underlying `OpenAL` with another library (e.g. RtAudio).

>You can also declare the defaults as constexpr and completely make them immutable, allowing modification only via code changes.

This also was my first thought, but it is just not the right thing to do here. Those values are really meant as defaults and thus mutable through public audio api.

>Another potential idea to reduce the size of the Audio class would be to declare a `AudioGainDevice`…

Yes, why not. I'd simply name that class `qTox::Audio::InputDevice` (including the namespaces). I think about splitting up audio devices for some time now, being able to actually use them when required and replace the "crappy" singleton and subscription mechanisms. OTOH, this PR is actually ready, making qTox "audio conferencing" finally controllable to users, which probably helps to extend the community (user base). I don't want to break that by bringing in new concepts too quickly. Hope you understand. Anyway I'd love to discuss future of qTox Audio API with a "professional attitude". ^^